cou/steps/hypervisor.py

```diff
@@ -104,15 +103,22 @@ def azs(self) -> list[HypervisorGroup]:
         """
         raise NotImplementedError
 
-    def _generate_pre_upgrade_plan(self) -> PreUpgradeStep:
+    def _generate_pre_upgrade_plan(self, target: OpenStackRelease) -> list[PreUpgradeStep]:
         """Generate pre upgrade plan for all applications.
 
-        This section should create a plan with steps like changing charm config option, etc.
+        This section should create a list of steps like changing charm config option, etc.
 
-        :return: Pre-upgrade step with all needed pre-upgrade steps.
-        :rtype: PreUpgradeStep
+        :param target: OpenStack codename to upgrade.
+        :type target: OpenStackRelease
+        :return: List of pre-upgrade steps.
+        :rtype: list[PreUpgradeStep]
         """
-        raise NotImplementedError
+        steps = []
+        for app in self.apps:
+            steps.extend(app.pre_upgrade_steps(target, units=None))
+            logger.info("generating pre-upgrade steps for %s app", app.name)
```

Comment:
I don't think this will work passing no units. There are some pre-upgrade steps, like refresh the current packages that needs to be targeted with the specific units in the group that it's going to be upgraded